README.md
# sigdb
[![Node.js CI](https://github.com/drazisil/sigdb/actions/workflows/node.js.yml/badge.svg)](https://github.com/drazisil/sigdb/actions/workflows/node.js.yml)
[![codecov](https://codecov.io/gh/drazisil/sigdb/branch/main/graph/badge.svg?token=3z7i3oZk67)](https://codecov.io/gh/drazisil/sigdb)

A file signature database manager

## Usage

To install:

`npm install -E sigdb`

To use:

```javascript
const { createSignatureDatabase }= require('sigdb')
const sigdb = await createSignatureDatabase()
```

## Methods

```javascript
  /**
   * Provided a buffer of binary bytes, will return either the matching signature name as a string, or `undefined`
   */
  find(haystack: Buffer): FileSignatureRecord | undefined
```

```javascript
  /**
   * Get an array of signature names in the database
   */
  getSignatureNames(): string[] 
```

```javascript
  /**
   * Add a file signature
   */
  addFileSignature(id: string, name: string, signature: Buffer): void 
```

## Types

```javascript
/**
 * @typedef FileSignatureRecord
 * @property {string} id
 * @property {Buffer} signature
 * @property {String} name
 */
```


CHANGELOG.md
# SigDB Changelog

## 3.0

### 3.0.1

Confirmed there is no breakage between the 3.x and 2.1.0 versions

### 3.0.0

Marking this a major in case it breaks. the API is the same, it starts life as typescript now.

* switch to typescript

## 2.0

We are starting here because this is the first version I'm taking notes and using semver on. Anything in the 1.0.x line is likely to have breaking changes from version to version. Sorry!

### 2.1.0

* add method for getSignatureNames
* add method for addFileSignature
* switch to typescript

tests/index.test.js
const { SigDB, loadSignatureDatabase } = require('../')
const tap = require('tap')

const sigDB = new SigDB()

tap.equal(
  sigDB.find(Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e])).name,
  'ar archive',
)

tap.test('Signature Database', async t => {
  const sigs = await loadSignatureDatabase()
  tap.equal(
    sigs.find(Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e])).name,
    'ar archive',
    'Can locate and match a file signature',
  )

  tap.equal(
    sigs.getSignatureNames().length,
    4,
    'Can get an array of file signatures',
  )

  sigs.addFileSignature('dummy1', 'dummySig', Buffer.alloc(3))
  tap.equal(sigs.getSignatureNames().length, 5, 'Can add a new file signature')
})


tests/index.test.ts
import { SigDB, loadSignatureDatabase } from '../lib//index'
import tap from 'tap'

const sigDB = new SigDB()

tap.equal(
  sigDB.find(Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e]))!.name,
  'ar archive',
)

tap.test('Signature Database', async t => {
  const sigs = loadSignatureDatabase()
  tap.equal(
    sigs.find(Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e]))!.name,
    'ar archive',
    'Can locate and match a file signature',
  )

  tap.equal(
    sigs.getSignatureNames().length,
    4,
    'Can get an array of file signatures',
  )

  sigs.addFileSignature('dummy1', 'dummySig', Buffer.alloc(3))
  tap.equal(sigs.getSignatureNames().length, 5, 'Can add a new file signature')
})


lib/index.ts
export interface FileSignatureRecord {
    id: string
    name: string
    signature: Buffer  
}

/** Class for manager file signatures (sometimes known as magic numbers) */
export class SigDB {
  private fileSignatures: FileSignatureRecord[]

  /** @deprecated please use loadSignatureDatabase() instead */
  constructor() {
    /** @type {FileSignatureRecord[]} */
    this.fileSignatures = [
      {
        id: '0',
        signature: Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e]),
        name: 'ar archive',
      },
      {
        id: '1',
        signature: Buffer.from([0xcf, 0xfa, 0xed, 0xfe]),
        name: 'mac executable',
      },
      {
        id: '2',
        signature: Buffer.from([0x7f, 0x45, 0x4c, 0x46]),
        name: 'elf executable',
      },
      {
        id: '3',
        signature: Buffer.from([0x4d, 0x5a]),
        name: 'dos mz executable',
      },
    ]
  }

  /**
   * Provided a buffer of binary bytes, will return either the matching signature name as a string, or `undefined`
   *
   * @param {Buffer} haystack
   * @memberof SigDB
   * @returns FileSignatureRecord | undefined
   */
  find(haystack: Buffer): FileSignatureRecord | undefined {
    return this.fileSignatures.find(sig => {
      return sig.signature.equals(haystack.slice(0, sig.signature.length))
    })
  }

  /**
   * Get an array of signature names in the database
   * @memberof SigDB
   * @returns {string[]}
   */
  getSignatureNames(): string[] {
    const results = []
    for (const entry of this.fileSignatures) {
      results.push(entry.name)
    }
    return results
  }

  /**
   * Add a file signature
   * @param {string} id 
   * @param {string} name 
   * @param {Buffer} signature 
   */
  addFileSignature(id: string, name: string, signature: Buffer): void {
    this.fileSignatures.push({id, name, signature})
  }
}

export function loadSignatureDatabase(): SigDB {
  return new SigDB()
}



dest/index.d.ts
/// <reference types="node" />
export interface FileSignatureRecord {
    id: string;
    name: string;
    signature: Buffer;
}
/** Class for manager file signatures (sometimes known as magic numbers) */
export declare class SigDB {
    private fileSignatures;
    /** @deprecated please use loadSignatureDatabase() instead */
    constructor();
    /**
     * Provided a buffer of binary bytes, will return either the matching signature name as a string, or `undefined`
     *
     * @param {Buffer} haystack
     * @memberof SigDB
     * @returns FileSignatureRecord | undefined
     */
    find(haystack: Buffer): FileSignatureRecord | undefined;
    /**
     * Get an array of signature names in the database
     * @memberof SigDB
     * @returns {string[]}
     */
    getSignatureNames(): string[];
    /**
     * Add a file signature
     * @param {string} id
     * @param {string} name
     * @param {Buffer} signature
     */
    addFileSignature(id: string, name: string, signature: Buffer): void;
}
export declare function loadSignatureDatabase(): SigDB;
//# sourceMappingURL=index.d.ts.map

dest/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadSignatureDatabase = exports.SigDB = void 0;
/** Class for manager file signatures (sometimes known as magic numbers) */
class SigDB {
    /** @deprecated please use loadSignatureDatabase() instead */
    constructor() {
        /** @type {FileSignatureRecord[]} */
        this.fileSignatures = [
            {
                id: '0',
                signature: Buffer.from([0x21, 0x3c, 0x61, 0x72, 0x63, 0x68, 0x3e]),
                name: 'ar archive',
            },
            {
                id: '1',
                signature: Buffer.from([0xcf, 0xfa, 0xed, 0xfe]),
                name: 'mac executable',
            },
            {
                id: '2',
                signature: Buffer.from([0x7f, 0x45, 0x4c, 0x46]),
                name: 'elf executable',
            },
            {
                id: '3',
                signature: Buffer.from([0x4d, 0x5a]),
                name: 'dos mz executable',
            },
        ];
    }
    /**
     * Provided a buffer of binary bytes, will return either the matching signature name as a string, or `undefined`
     *
     * @param {Buffer} haystack
     * @memberof SigDB
     * @returns FileSignatureRecord | undefined
     */
    find(haystack) {
        return this.fileSignatures.find(sig => {
            return sig.signature.equals(haystack.slice(0, sig.signature.length));
        });
    }
    /**
     * Get an array of signature names in the database
     * @memberof SigDB
     * @returns {string[]}
     */
    getSignatureNames() {
        const results = [];
        for (const entry of this.fileSignatures) {
            results.push(entry.name);
        }
        return results;
    }
    /**
     * Add a file signature
     * @param {string} id
     * @param {string} name
     * @param {Buffer} signature
     */
    addFileSignature(id, name, signature) {
        this.fileSignatures.push({ id, name, signature });
    }
}
exports.SigDB = SigDB;
function loadSignatureDatabase() {
    return new SigDB();
}
exports.loadSignatureDatabase = loadSignatureDatabase;
//# sourceMappingURL=index.js.map